src/core/QCheck_ng.ml

```diff
@@ -0,0 +1,1956 @@
+(*
```

Comment:
Moving the discussion to a comment for thread-like features and a better organization of discussions.

@c-cube said:
> *edit*: actually, it could even be done all from within `QCheck.ml`, if we introduce `GenShrink` or `GenTree`, keep the old `Gen` and `Shrink` (and flag them as deprecated when the new API is stable enough). Then we can have:

```ocaml
type 'a arbitrary =
  | Old of { 'a gen; 'a shrink; 'a print; small }
  | New of { 'a gentree; 'a print }
```

> (hiding the definition in the .mli, which should never have been exposed. Lessons learnt on API design.) Then current code still just works, albeit with deprecation warnings in the future; and people can migrate incrementally.

I have mixed feelings about it on second thought because all functions that manipulate more than 1 arbitrary (e.g. `pair`) or functions that manipulate `Gen.t` too (e.g. `array_of_size`) will lead to weird and maybe even impossible code.

Considering `pair (Old g1) (New g2)`, what shrinking function do we use? Do we use `gen` or `gentree`?
Remember also that the running code (like type `state` or function `check_state_input` had to be changed to use the new shrinking code.

Maybe unification can happen higher (e.g. the QCheck runner) but I think it's just a lot easier to keep `QCheck_ng` separated

Reply:
Maybe a naive question but could you not defining `{ 'a gen; 'a shrink; 'a print; small }` in term of `'a gentree`?

Reply:
Sooooo after quite some work these last 2 days, I think the next (and last unknown) problem to tackle is this (and then it should be simple code deletion, documentation cleanup/proofreading).

I'm posting this message both to discuss the matter (and get ideas) and as a thinking support for myself.

It's not that simple to have a solution that suits all of:
1. code still compiles and works when users update the QCheck version in their deps
2. code can be migrated little by little to QCheck2
3. No code duplication

## Runners
All the runners (Alcotest, OUnit, Raw, Base) currently use explicitly `QCheck.XXX` in their signatures

Options:
* Copy/paste them to `QCheck2_alcotest`, etc.
  * :-1: Code duplication
  * :-1: Migration must happen module by module
  * :+1: Code compiles after upgrade
* Define a module signature that both `QCheck` and `QCheck2` are compatible with, and make all runners functors
  * :+1: No code duplication
  * :-1: Migration must happen module by module
  * :-1: Code doesn't compile after upgrade
* Same but all runners include `Functor (QCheck)`
  * :+1: No code duplication
  * :-1: Migration must happen module by module
  * :+1: Code compiles after upgrade
  * :-1: QCheck2 feels like a second class citizen, users must write `module QCheck2_alcotest = QCheck_alcotest.Make (QCheck2)
* Do some Dune magic (is that even possible? I am invoking @c-cube's or anyone else's knowledge on the matter :grin: ) so that only the functors applied to `QCheck` and `QCheck2` are exported
  * :+1: No code duplication
  * :-1: Migration must happen module by module
  * :+1: Code compiles after upgrade
  * :+1: Each runner lib exports both `QCheck_xxx` and `QCheck2_xxx` modules

## Arbitraries
@gr-im makes a valid point: it is probably possible to convert a QCheck.arbitrary to a QCheck2.arbitrary (but not the other way around).

So we can provide a function `QCheck2.migrate_from_v1_arbitrary (arb : QCheck.arbitrary) : QCheck2.arbitrary` to help users migrate arbitraries little by little to QCheck2 (with a deprecation warning) if it helps some users.

Implementation wise, this should be approximately:
```ocaml
let {QCheck.gen = gen_v1; shrink = shrink_v1_opt; ...} = arb in
let shrink = match shrink_v1_opt with
  | None -> fun _ -> Seq.empty
  | Some shrink_v1 -> fun x ->
    let rev_shrinks = ref [] in
    shrink_v1 x (fun y -> rev_shrinks <- y :: !rev_shrinks) ;
    List.rev rev_shrinks |> List.to_seq
in
let gen = Gen.make_primitive gen_v1 shrink in
{gen; ...}
```

## Tests
Similarly I think we can provide a function `QCheck2.migrate_from_v1_test` to migrate a `QCheck.Test.t` to a `QCheck2.Test.t` (the only moving part is the arbitrary, we would only have to expose getters for all fields of `QCheck.Test.t` as it is opaque) by relying on `migrate_from_v1_arbitrary` mentioned above.

Thus users **could** run together (using the QCheck2 runner) a mix of QCheck tests (by wrapping them with `migrate_from_v1_test` and QCheck2 tests.

Reply:
FYI I am currently exploring an idea:
* `QCheck_base_runner_internal.ml[i]` that contains basically everything the current `QCheck_base_runner` contains, wrapped in a functor that takes as argument a module with the signature of `QCheck[2]` (minimalist)
* `QCheck_base_runner.ml[i]` is only the functor applied to `QCheck`
* `QCheck2_base_runner.ml[i]` is only the functor applied to `QCheck2`

I feel like this is the best option right now:
* users that update the QCheck dep version will not see a diff
* when users want to start using QCheck2, they must also switch the runners to the `2` flavor and wrap all tests/arbitraries with the `migrate` functions (maybe the `QCheck2_base_runner` can additionnally provide `migrate` wrapping functions too to simplify?)

I am still young in the module land of OCaml so I don't have enough experience to "feel" if this is a good idea/solution for now, input welcome :bow:

Reply:
Ok I pushed the result for `QCheck_base_runner`. As you can see, there are now both `QCheck_base_runner` and `QCheck2_base_runner`

@c-cube are you ok with this solution? If so I can do the same for OUint and Alcotest and then we should be good to go.

Reply:
I wish we could find a solution where tests from QCheck and from QCheck2 can be mixed freely. Maybe the `QCheck.Test.t` part could be extracted into a separate module (`QCheck_test`?) and use an abstract notion of generator/arbitrary? It could take something like:

```
module type ARB = sig
  type generated
  type elt

  val prop : elt -> bool
  val gen : Random.State.t -> generated
  val cur_elt : generated -> elt
  val shrink : generated -> generated Seq.t

  val pp : elt printer
end

type 'a arb = (module ARB with type elt = 'a)
```

and then QCheck2 has: `generated = elt Tree.t` and QCheck has `generated = elt`. Do you think that's doable? It'd make migration of test suites much easier.

Reply:
Just to be safe: have you seen I have written a `migrate_v1_test` function, and you don't think this is good enough? That means currently one can almost transparently mix v1 and v2 tests, up to wrapping the v1 tests with this function (worste case scenario, a `List.map migrate_v1_test`).

I don't mind investigating this idea if you dislike the wrapping function, I just want to clear that out :sweat_smile:

Reply:
I think a migration function is an ok solution if we can't find better, but I do hope we can mix them more seamlessly (and not duplicate all runners). :slightly_smiling_face:

Reply:
@c-cube asked me to share my ideas on this. I haven't been following the PR, so this may all fill a bit raw and out of context.

The bottom-line is that it is indeed possible to make a type which instantiate to both the two sort of arbitraries.

### Where we are at

Let me establish some terminology here, I'll call the types `Old.arbitrary` and `New.arbitrary` for the obvious reasons.

Up to details which don't concern us today, we have

```ocaml
type 'a Old.arbitrary = {
  gen : 'a Gen.t
  shrink : 'a -> 'a List.t
}

type 'a New.arbitrary = {
  gen : 'a Tree.t Gen.t
}
```

### Generalising

Now, let us define the more general `General.arbitrary` which instantiate into both:

```ocaml
type ('x,'a) General.arbitrary = {
  gen : 'x Gen.t
  display : 'x -> 'a
  shrink : 'x -> 'x List.t
}
```

Now we have

```ocaml
type 'a Old.arbitrary = ('a,'a) General.arbitrary
```

With `display x = x`. And


```ocaml
type 'a Old.arbitrary = ('a Tree.t, 'a) General.arbitrary
```

With `display = Tree.root` and `shrink = Tree.children` (or whatever you want to call the fields of a rose tree's nodes).

### Defining tests

Now, we can easily define tests to take an `('x,'a) General.arbitrary` (please find a better name for it…) instead of an `'a Old.arbitrary`: the test generates `'x`, converts them into `'a` for testing, shrinks `'x`. It should be a rather straightforward generalisation.

Crucially, we don't want to change the `'a cell` type. The function `make` becomes

```ocaml
make : ('x,'a) arbitrary -> ('a -> bool) -> 'a cell
```

There is some existential quantification going on somewhere, here. But the key point is that we now have a single type `'a cell` of tests using either a old-style or a new-style `arbitrary`.

Mission accomplished, I think.

### Consideration about backward compatibility

Prior to this PR, the type `arbitrary` is a private record (rather than an abstract type). An interesting consequence of this here proposal is that it simply adds one field to the record: it's a backward compatible change.

That being said, even if, ultimately, you decide that it is better for `arbitrary` to be abstract, this proposal still have the property of exposing the two kinds of arbitrary to be instances of one more general type, which I think will be the better path to transitioning.

### Theoretical considerations

It's worth noting that the impredicative encoding of the type of rose trees (_grumble something about co-inductive types here_) is `∃'x. 'x × ('x → 'a) × ('x → 'x List.t)`. This is almost the same type as `∃'x. ('x, 'a) General.arbitrary` which simply adds a `Gen.t` around the `gen` field.

We've already seen how a `'a New.arbitrary` could be seen as an `∃'x. ('x,'a) General.arbitrary`. Conversely, the `map` function of `Gen.t` let's us convert an `∃'x. ('x,'a) General.arbitrary` in a `(∃'x. 'x × ('x → 'a) × ('x → 'x List.t)) Gen.t = 'a New.arbitrary` (_something, something, functorial strength, something_). Is suspect this is some kind of isomorphism. Not much information seems to lost when transforming in either direction, and it's probably irrelevant to our application.

This indicates that the new-style arbitrary are indeed most general in a certain sense. The strength of this here proposal is only to have the types align well.

As a final consideration (which may be the same remark, really), let me point out that `∃'x. ('x,'a) General.arbitrary` has a natural applicative functor structure (exercise left to the reader 🙂). I believe this structure to be the same as the structure that Hedgehog and @Sir4ur0n gave to `New.arbitrary`. So it ought to be possible to define directly `type 'a New.arbitrary = ∃'x. ('x, 'a) General.arbitrary`. That would spare the introduction of the Rose tree type. But I think that, from the point of view of unifying old tests and new tests, this would not work quite as well.

Reply:
Thank you for your input!

With this solution, we couldn't keep `Old.arbitrary` private (backward compatibility) while making `New.arbitrary` abstract, could we? IIUC if we make `New.arbitrary` abstract, it won't be part of the interface that `'a New.arbitrary = ('a Tree.t, 'a) General.arbitrary` and thus we won't be able to have an heterogeneous list of QCheck/QCheck2 tests.

I am also slightly reluctant in introducing yet another indirection layer for backward compatibility, but maybe this is just me :sweat_smile: 

In parallel I have explored another idea: Remove `QCheck.Test/TestResult` and have `QCheck.Test.make[_cell]` return QCheck2 types instead, since I think test developers are solely interested in those 2 functions, others are for runners.

* runners all run lists of `QCheck2.Test.t` (no duplicate runners)
* `QCheck.Test.make[_cell]` returns QCheck2 tests
* Obviously `QCheck2.Test.make[_cell]` return QCheck2 values :grin: 

You can see the result (unclean, a lot of code commented out) in https://github.com/Sir4ur0n/qcheck/commit/4aadcd131b0086b0565b678ecbbcc11e0698d57f (not pushed on this branch yet to avoid back-and-forths on commits).

I tested this on an existing codebase using QCheck and it works OOTB (up to the change of type `QCheck.Test.t` -> `QCheck2.Test.t` as we have helper functions)

Reply:
With the approach I outline, you would have `General.arbitrary` be the private type (or abstract if you are so inclined) and both `New.arbitrary` and `Old.arbitrary` be manifest types.

This way you have a single `make_cell` function which takes either kind of generators.

Reply:
After discussing in private with @aspiwack: it is indeed impossible to have at the same time `QCheck2.arbitrary` abstract and `QCheck.arbitrary` private with this solution

@c-cube That being said, I guess it's your call now:
* do we go for "Both generate QCheck2 tests"? Cf https://github.com/Sir4ur0n/qcheck/commit/4aadcd131b0086b0565b678ecbbcc11e0698d57f 
* do we go for "One type to rule them all, but make `QCheck2.arbitrary` private instead of abstract"? Cf Arnaud's proposal
* another solution?

Reply:
> I'm against breaking all existing code by requiring an explicit migrate_to_v2 function.

I don't know if you are saying this in response to the first option, but: the solution I linked in my previous message doesn't require users to call `migrate` functions (it's conceptually the opposite: `QCheck.Test.make` calls `migrate` under the hood to return a `QCheck2.Test.t`). It's transparent to the user (except this now returns a `QCheck2.Test.t` instead of a `QCheck.Test.t`). The `migrate` functions are no longer in the API.

Reply:
I think the linked commit showed a lot of commented code except for the migration function :slightly_smiling_face: .

Reply:
I just checked again and the migration code is indeed commented out, isn't it?

* [QCheck2.mli](https://github.com/Sir4ur0n/qcheck/commit/4aadcd131b0086b0565b678ecbbcc11e0698d57f#diff-8fd067287338571071bfdd73627f8cde2084e48fcf5abf02f54f2b56f19a5405R1589-R1605)
  * `migrate_v1_arbitrary`
  * `migrate_v1_cell`
  * `migrate_v1_test`
* [QCheck2.ml](https://github.com/Sir4ur0n/qcheck/commit/4aadcd131b0086b0565b678ecbbcc11e0698d57f#diff-07dbdff6f84a44f565ad4871c5790ad598ceda123f0b8190a711d303d3be28efR2029-R2056) where I commented out

If you're ok with this, I will cleanup the code.

By the way is there a trick to build the `qcheck-ounit` sub-package? `make` at the root succeeds even when this package is obviously broken

Reply:
You need to have ounit2 installed I think. Dune doesn't build packaged whose dependencies are not satisfied.

Alternatively, `dune build -p qcheck-ounit` might work…

Reply:
Thank you, indeed `opam install ounit2` was the solution :+1: 

I have cleaned up the commented code and pushed back the `no-duplicate` branch on the current PR branch, let me know what you think